spec.md

```diff
@@ -2563,46 +2563,89 @@ Compiler _implementation_.  A FIRRTL Compiler is a program that converts FIRRTL
+scalarized convention should be the default convention for "public" modules,
+such as the top module of a circuit, "device under test", or an ext module.
 
-The Lower Types algorithm operates as follows:
+The lowering algorithm for the scalarized convention operates as follows:
 
-1. Ground type names are unmodified.
+1. Ports are scalarized top-to-bottom.
 
-2. Vector types are converted to ground types by appending a suffix, `_<i>`, to
-   the i^th^ element of the vector.  (`<` and `>` are not included in the
-   suffix.)
+2. Ground-typed ports names are unmodified.
 
-3. Bundle types are converted to ground types by appending a suffix, `_<name>`,
-   to the field called `name`.  (`<` and `>` are not included in the suffix.)
+3. Vector-typed ports are scalarized to ground-typed ports by appending a
+   suffix, `_<i>`, to the i^th^ element of the vector. Elements are scalarized
+   recursively, depth-first, and left-to-right.
 
-New names generated by Lower Types must be unique with respect to the current
-namespace (see [@sec:namespaces]).
+4. Bundle-typed ports are scalarized to ground-typed ports by appending a
+   suffix, `_<name>`, to the field called `name`. fields are scalarized
+   recursively, depth-first, and left-to-right.
 
-E.g., consider the following wire:
+E.g., consider the following port:
 
 ``` firrtl
-wire a : { b: UInt<1>, c: UInt<2> }[2]
+module Top :
+  input a : { b: UInt<1>, c: UInt<2> }[2]
 ```
 
-The result of a Lower Types pass applied to this wire is:
+Scalarization breaks `a` into the following ports:
 
 ``` firrtl
-wire a_0_b : UInt<1>
-wire a_0_c : UInt<2>
-wire a_1_b : UInt<1>
-wire a_1_c : UInt<2>
+module Top :
+  input a_0_b : UInt<1>  ; a[0].b
+  input a_0_c : UInt<2>  ; a[0].c
+  input a_1_b : UInt<1>  ; a[1].b
+  input a_1_c : UInt<2>  ; a[1].c
 ```
 
+The body of a module defininition introduces a new, empty namespace. As new
+port names are added, these names must be unique with respect to this
+namespace. In the case of a collision during renaming, priority will be given
+to values that are converted first.
+
+If a name is already taken, that name will be made unique by appending a suffix
+`_<i>` to the name, where `i` is the lowest nonnegative integer that gives a
+unique name.
+
+E.G. consider the following ports:
+
+```
+module Top :
+  input a : { b: UInt<1>[2], b_0: UInt<2>, b_1: UInt<3>}
+  input a_b: UInt<4>[2]
+  input a_b_0: UInt<5>
+```
+
+Scalarization breaks these ports into the following ports:
+
+```
+module Top :
+  input a_b_0: UInt<1>    ; a.b[0]
+  input a_b_1: UInt<1>    ; a.b[1]
+  input a_b_0_0: UInt<2>  ; a.b_0
+  input a_b_1_0: UInt<3>  ; a.b_1
+  input a_b_0_1: UInt<4>  ; a_b[0]
+  input a_b_1_1: UInt<4>  ; a_b[1]
+  input a_b_0_2: UInt<5>  ; a_b_0
+```
+
+Named components in the body of a module will be renamed as needed to ensure
+port names follow this convention.
+
+## The "Internal" Convention
+
+Private modules (i.e. modules that are _not_ the top of a circuit, device under
```

Comment:
Do we want to define public/private directly somewhere?  As written, the language here uses the concept and clarifies what is meant but only indirectly defines these terms.  Not sure where this would go, perhaps we tackle that separately.

Reply:
yeah i think it would be good to separate out the 'public/private' definition from the definition of these representations.

Reply:
I agree, but will push that out to a different PR, it that's okay.

Reply:
Absolutely! This is tracked in #82 already, all good.  Thanks!